### PR TITLE
Add copy codeblock button to built-in documentation

### PR DESCRIPTION
**Pull request by Mickeon**
Would combine really well with #87301

This PR adds a Copy button to the codeblock examples in the built-in class reference.

Unfortunately, with the icon, the underline still shows up and cuts through the icon. This is a bug with RichTextLabel itself.
image

If the underline [could be removed outright](https://github.com/godotengine/godot/pull/64861) this would be less of an issue, but alas. So in the current PR it looks like this:
image

### CODE REVIEW

File: editor/editor_help.cpp [hunk 1]
```diff
@@ -2097,18 +2103,26 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, Control
...
 				} else if (codeblock_tag) {
 					p_rt->pop(); // font size
-					// Pop color, cell and table.
-					p_rt->pop();
-					p_rt->pop();
-					p_rt->pop();
+
+					// Copy codeblock button.
+					p_rt->push_cell();
+					p_rt->set_cell_row_background_color(code_bg_color, code_bg_color);
```
Comment (YuriSizov) [inline editor/editor_help.cpp:2116]: This doesn't use exactly the same color for the second argument. For some reason we set up the first cell with the even color being at 0.99 alpha. Not sure why it's relevant, but we should change either that line or your addition, so they are the same.

File: editor/editor_help.cpp [hunk 2]
```diff
@@ -2097,18 +2103,26 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, Control
...
+					p_rt->set_cell_padding(Rect2(0, 10 * EDSCALE, 0, 10 * EDSCALE));
+					p_rt->push_meta("^" + codeblock_text);
+					codeblock_text = "";
+					p_rt->append_text(TTR("Copy", "Used in the built-in documentation's codeblocks"));
+					// p_rt->add_image(p_owner_node->get_editor_theme_icon(SNAME("ActionCopy")), 32 * EDSCALE, 32 * EDSCALE, Color(1, 1, 1, 0.3), INLINE_ALIGNMENT_CENTER, Rect2(), Variant(), true, TTR("Click to copy"));
```
Comment (YuriSizov) [inline editor/editor_help.cpp:2121]: You should probably remove it if we can't use it right now.

File: editor/editor_help.cpp [hunk 3]
```diff
@@ -2097,18 +2103,26 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, Control
...
+					p_rt->pop(); // meta
+					p_rt->pop(); // cell
+
+					p_rt->pop(); // table
```
[the remark this answers is not in the document]
Comment (Mickeon) [reply editor/editor_help.cpp:2125]: I spent so long figuring out why and I still have no idea. For some reason one less `pop` for the previous cell does **not** break. Defining a `pop` right before pushing the new cell completely breaks everything.

**Review by YuriSizov — CHANGES_REQUESTED** [review]
It doesn't quite work at the moment. Left-brackets get eaten somewhere


https://github.com/godotengine/godot/assets/11782833/3532250c-15f0-42b3-9b84-4f23839130e5

It would also be nice to add some visual indication to the click. We might need to modify RTL to achieve that, but without any feedback it feels very poor.

On the subject of modifying RTL, we should be able to add a new argument to the `meta` tag which allows disabling the underlining on the case by case to solve your issue with the icon.

File: editor/editor_help.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (YuriSizov) [reply editor/editor_help.cpp:2125]: You need 2 pops before the new cell. One for color, one for cell.

**Comment by Calinou** [comment]
> It would also be nice to add some visual indication to the click. We might need to modify RTL to achieve that, but without any feedback it feels very poor.

I've definitely been missing hover/pressed feedback in RTL (also in projects), so we'll probably need to modify RTL to support this using additional theme items.

File: editor/editor_help.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Mickeon) [reply editor/editor_help.cpp:2125]: I understood and reorganised that whole code because it was mislabeled.

**Comment by Mickeon** [comment]
The PR has been updated to fix the above quirks. Still, it cannot be merged as is because the above things are missing in RichTextLabel.

File: editor/editor_help.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (YuriSizov) [reply editor/editor_help.cpp]: Try this:

```
diff --git a/editor/editor_help.cpp b/editor/editor_help.cpp
index 85b810d832..915c2cbaac 100644
--- a/editor/editor_help.cpp
+++ b/editor/editor_help.cpp
@@ -2105,13 +2105,14 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, Control
 			tag_stack.pop_front();
 			pos = brk_end + 1;
 			if (tag != "/img") {
-				p_rt->pop(); // font
 				if (code_tag) {
-					p_rt->pop(); // font size
 					p_rt->pop(); // color
 					p_rt->pop(); // background color
-				} else if (codeblock_tag) {
 					p_rt->pop(); // font size
+					p_rt->pop(); // font
+				} else if (codeblock_tag) {
+					p_rt->pop(); // color
+					p_rt->pop(); // cell
 
 					// Copy codeblock button.
 					p_rt->push_cell();
@@ -2125,6 +2126,10 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, Control
 					p_rt->pop(); // cell
 
 					p_rt->pop(); // table
+					p_rt->pop(); // font size
+					p_rt->pop(); // font
+				} else {
+					p_rt->pop(); // font
 				}
 			}
 			code_tag = false;

```

[the remark this answers is not in the document]
Comment (Mickeon) [reply editor/editor_help.cpp]: A little too late 💦 Should I prefer this snippet of code over the current one?

[the remark this answers is not in the document]
Comment (YuriSizov) [reply editor/editor_help.cpp]: You did the same exact thing, all good :)

**Comment by Mickeon** [comment]
Sidenote, all of these desired features may give more "credence" to implement https://github.com/godotengine/godot-proposals/issues/7617, but that would be a huge undertaking.

**Review by YuriSizov — APPROVED** [review]
Seems to work correctly now. However, without https://github.com/godotengine/godot/pull/87301, or something similar, this isn't really practical. Copied code snippets will immediate err and throw the user into confusion.

**So approving for functionality, but as is, I wouldn't merge it.** Usability improvements discussed above would also be nice (can be implemented separately as a pre-requisite, then this could be updated on top).

**Comment by KoBeWi** [comment]
Needs rebase.

**Comment by Mickeon** [comment]
I will gladly rebase if this is desired, even without the above omissions due to RichTextLabel's limitations

File: editor/editor_help.cpp [hunk 4]
```diff
@@ -2384,6 +2386,7 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, Control
 	List<String> tag_stack;
 	bool code_tag = false;
 	bool codeblock_tag = false;
+	String codeblock_text = "";
```
[the remark this answers is not in the document]
Comment (AThousandShips) [reply editor/editor_help.cpp:2389]:
```suggestion
	String codeblock_text;
```
At least no unnecessary default value

**Review by KoBeWi — APPROVED** [review]
Looks fine. The Copy button doesn't overlap anything, so as long as it's usable, it can be improved later.

The `codeblock_text` optimization I mentioned should be looked into, as appending that many strings might actually make the help slower to load (though it doesn't seem to have significant impact).

**Comment by Mickeon** [comment]
The entire built-in class ref generation would benefit from StringBuilder _t.b.h._. I don't know what that is, never used it, but maybe it'll be the occasion to...

For now I updated the PR to remove the default `""` value but maybe I can figure this one out before a merge.

**Comment by Mickeon** [comment]
Hey, that's not too bad. Updated the PR to use StringBuilder. This "thing" could be used in so many other places...

File: editor/editor_help.cpp [hunk 5]
```diff
@@ -2384,6 +2387,7 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, Control
 	List<String> tag_stack;
 	bool code_tag = false;
 	bool codeblock_tag = false;
+	StringBuilder codeblock_text; // More efficient to append with.
```
Comment (KoBeWi) [inline editor/editor_help.cpp:2390]: I think the comment is not necessary, as StringBuilder's purpose is self-explanatory.

Comment (Mickeon) [reply editor/editor_help.cpp]: Just like that gone in a flash.

**Comment by Mickeon** [comment]
I can do that. There is still a problem of "lack of feedback" when hovering and clicking over the button that the "Copy" text would mitigate (because the underline could appear on hover).

**Comment by Mickeon** [comment]
I updated the PR to use an image. If it is settled to use raw text, we can go back on it just fine.

At first I went with this but I thought it was _way_ too highlighted.
image

So I settled with something like this:


You may notice that the underline is a bit bugged out when hovering on the side, but that's RichTextLabel's fault (again).

The gif itself is outdated as I made the icon ever so slightly smaller.
image

**Comment by akien-mga** [comment]
The icon looks a bit blurry to me, is there a scaling issue?

**Comment by Mickeon** [comment]
Yeah. It's 24x24 (excluding editor scale), I can tone the scale back to 16x16 (as per the previews in the description of the PR) if you wish.

**Comment by bruvzg** [comment]
> You may notice that the underline is a bit bugged out when hovering on the side, but that's RichTextLabel's fault (again). 

Seems like meta hover callbacks are using wrong argument (for selection, instead of meta), will fix it in a moment.

**Comment by Mickeon** [comment]
Definitely out of scope here. The "blurriness" has been an issue with other icons in the built-in documentation, as well. It's just more noticeable here because it is larger than usual.
image

So, should I reduce the size until the issue with blurry icons is resolved?

**Comment by akien-mga** [comment]
I guess it's fine as is for now, but we really need to solve these blurry icons :)

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**